### PR TITLE
161780568 user  can like/dislike articles

### PR DESCRIPTION
**Pull request by innocentEdosa**
#### Description
This PR sets up user can like and dislike article functionality.

#### Type of change
- Feature 
 - create like controller
 - setup like model and migration
 - setup like route
 - write test for like functionality

#### How Has This Been Tested?
clone the repo
git chekout to feature/161780568-user-like/dislike-articles
run npm start
test endpoint with postman

#### Pivotal tracker id
161780568

#### Flow chart
Blank Diagram

#### Postman Tests

- Liking an article
Screenshot 2019-04-12 at 6 29 29 PM

- Unliking an article
Screenshot 2019-04-12 at 6 29 35 PM

- Disliking an article
Screenshot 2019-04-12 at 6 29 46 PM

### CODE REVIEW

File: src/routes/api/articles/likeController.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      }
+      await Likes.changeOption(like, option);
+      await Likes.countOptions(request, response, option);
+    } catch (err) {
+      return next(err);
+    }
+  }
+
+  /**
+ *
```
Comment (kossy360) [inline src/routes/api/articles/likeController.js]: add a description of what this method does

Comment (innocentEdosa) [reply src/routes/api/articles/likeController.js]: Thank you Kossy for the review
However, I do not fully understand what you want me to do.
Please kindly explain

Thanks

File: src/routes/api/articles/likeController.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * handles functionality for toggling like and
+ * dislike of an article
+ * @static
+ * @param {Request} request
+ * @param {Response} response
+ * @param {next} next
+ * @memberof Articles
+ * @returns {void}
+ */
+  static async likeArticle(request, response, next) {
```
Comment (kossy360) [inline src/routes/api/articles/likeController.js]: method name should describe what it does

File: src/routes/api/articles/articleRoutes.js [hunk 3]
```diff
@@ -3,14 +3,17 @@ import validate from '../../../middlewares/validator/index';
...
 const { validateArticle, validateArticleSearch } = validate;
 const { createArticle } = articleController;
-
+const { likeArticle } = likeController;
 const articleRoute = Router();
 
 articleRoute.post('/', authenticator, validateArticle, createArticle);
 
 articleRoute.get('/search', validateArticleSearch, searchArticles);
+articleRoute.post('/:slug/:option', authenticator, likeArticle);
```
Comment (kossy360) [inline src/routes/api/articles/articleRoutes.js]: add validation for request parameters

File: src/routes/api/articles/likeController.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    try {
+      let option = true;
+      if (request.params.option === 'dislike') { option = false; }
```
Comment (kossy360) [inline src/routes/api/articles/likeController.js]: an API call with defined parameters should only perform a single action. modify controller to accept third parameter for clearing a like or dislike

**Review by kossy360 — DISMISSED** [review]

File: src/routes/api/articles/likeController.js (hunk 4 above)
Comment (innocentEdosa) [reply src/routes/api/articles/likeController.js]: Thank you Kossy360 for the review.

However, this controller handles all functionality relating to article likes and dislikes. A user can like & clear their like or dislike & clear their dislike. It basically acts as a toggle. In the end, it performs a single action which is handling like-dislike functionality.

Thanks

File: src/middlewares/validator/index.js [hunk 5]
```diff
@@ -109,6 +109,23 @@ class Validate {
         status: 422, error,
       }));
   }
+
+  /**
+ *
```
Comment (kossy360) [inline src/middlewares/validator/index.js]: add method description

File: src/routes/api/articles/likeController.js (hunk 4 above)
Comment (kossy360) [reply src/routes/api/articles/likeController.js]: To clarify, an API call in the format ```{baseUrl}/articles/<slug>/like``` should only have one responsibility; liking an article. It should not toggle between liking and clearing a like.
For the extra functionality of clearing a like or dislike, I would advice creating a request like so ```{baseUrl}/articles/<slug>/clearlike``` to clear either a like or a dislike

**Comment by kossy360** [comment]
PR has two commits, please squash.

File: src/routes/api/articles/articleRoutes.js (hunk 3 above)
Comment (innocentEdosa) [reply src/routes/api/articles/articleRoutes.js]: handled

File: src/routes/api/articles/likeController.js (hunk 2 above)
Comment (innocentEdosa) [reply src/routes/api/articles/likeController.js]: I just changed the name to likeArticleHandler, I feel like should be a general for likes and dislike
but which name would you consider appropriate?

Thanks

File: src/routes/api/articles/likeController.js (hunk 4 above)
Comment (innocentEdosa) [reply src/routes/api/articles/likeController.js]: Noted, Thanks

File: src/middlewares/validator/index.js (hunk 5 above)
Comment (innocentEdosa) [reply src/middlewares/validator/index.js]: Done now

Thanks

**Review by profmcdan — CHANGES_REQUESTED** [review]
Kindly review your implementation of countOption and the handler, especially how they depend on each other.

**Review by profmcdan — APPROVED** [review]

**Pull request merged**